src/extensions/extensionsHelper.ts

```diff
@@ -479,51 +568,78 @@ export async function incrementPrereleaseVersion(
 }
 
 /**
- * Validates the Extension spec.
+ * Gets a list of the next version to upload by release stage.
  *
- * @param publisherId the ID of the Publisher
- * @param extensionId the ID of the Extension
- * @param rootDirectory the directory with the Extension's source
- * @param latestVersion the latest version of the Extension (if any)
- * @param stage the release stage
+ * @param extensionRef the ref of the extension
+ * @param version the new version of the extension
+ */
+async function getNextVersionByStage(
+  extensionRef: string,
+  newVersion: string
+): Promise<{ versionByStage: Map<string, string>; hasVersions: boolean }> {
+  const versionByStage = new Map(
+    ["rc", "alpha", "beta"].map((stage) => [
+      stage,
+      semver.inc(`${newVersion}-${stage}`, "prerelease", undefined, stage)!,
+    ])
+  );
+  let extensionVersions: ExtensionVersion[] = [];
+  try {
```

Comment:
This function is giving me some bad code smells - right now, it is very error driven, and throws+catches errors during the normal & intended behavior of handling stable versions.
Could we avoid this if we filter out all ev's that don't have a prerelease?

Reply:
Or, alternatively, we should error out if a stable version with this id already exists, since the upload will definitely fail then.

Reply:
I agree it kinda sucks but it's limited by the API. Also, the try/catch is only there to handle 404s, not for anything else. I can refactor to make the try/catch around the API fetch only.

The stable is outside of the main logic because regardless of what exists already, the stable version pulled from extension.yaml is what would be uploaded next (whether it succeeds or is valid or not).

We do error out prior to this in `validateVersion()` if `newVersion` is less than or equal to the latest version.

Reply:
Refactored it to hopefully make it a bit clearer/less sketchy looking.